Fix `OptimizeSwapBeforeMeasure` crash when two measures share a clbit

p51lee opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

**Fixes #14949**

`OptimizeSwapBeforeMeasure` followed classical descendants from a `swap` and tried to modify a later `Measure` that only reused the same clbit (it did not measure either swapped qubit), raising: `ValueError: tuple.index(x): x not in tuple`.

### Details and comments

This change modifies a `Measure` node only when its `qargs[0]` is one of the swap qubits. Unrelated measures are left untouched. The measure layer is rebuilt in topological order to keep write order to the shared clbit.

New regression test fails on `main` and passes with this fix.

p51lee committed:
Fix `OptimizeSwapBeforeMeasure` transpiler pass crash

Fix crash when two measures write to the same clbit.

Only modify measures that read one of the swapped qubits; leave unrelated
measures untouched. Rebuild the measure layer in topological order.

Refs: #14949

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14950) <br/>All committers have signed the CLA.

p51lee committed:
Merge branch 'main' into fix/opt-swap-measure-clbit-reuse

jakelishman commented on qiskit/transpiler/passes/optimization/optimize_swap_before_measure.py:
```diff
@@ -56,16 +57,24 @@ def run(self, dag):
+                    successors = swap_descendants
+                else:
+                    # following the topological order
+                    successors = (s for s in dag.topological_op_nodes() if s in swap_descendants)
```
I'm a bit concerned about the computational complexity of this - we're now iterating through the entire set of topological nodes for every swap with more than one descendant.

Keeping this spirit of this PR, perhaps a solution is to produce something like
```python
order = {node: i for i, node in enumerate(dag.topological_nodes())}
```
at the top of the pass, then do the later ordering as `sorted(swap_descendents, key=order.get)`?

p51lee replied:
I agree—computing the order only once would be better, but DAG mutations make that tricky.

I first tried precomputing `{node -> idx}` once at the beginning, but since the main loop mutates the DAG (removes swap/changes measures), that snapshot becomes stale—the node set/order changes. I also tried deferring mutations out of the main loop, but each iteration's measure descendants depend on the previous rewrite, so computing everything from the initial graph gives the wrong result.

For a slight improvement I plan to early-filter and only sorting when it matters:
```python
measures = [
    successor for successor in swap_descendants
    if isinstance(successor, DAGOpNode) and isinstance(successor.op, Measure)
]
if len(measures) > 1:
    order = {node: i for i, node in enumerate(dag.topological_nodes())}
    measures.sort(key=order.get)
for successor in measures:
    # (REMOVED) if isinstance(successor, DAGOpNode) and isinstance(successor.op, Measure):
        ...
```
This avoids a full DAG scan in the common case (each qubit/clbit measureed once). If you have a preffered solution, I'm happy to switch.

jakelishman commented on releasenotes/notes/opt-swap-measure-clbit-reuse-938bab7e1968bea5.yaml:
```diff
@@ -0,0 +1,7 @@
+fixes:
+  - |
+    **Transpiler:** Fix a crash in
```
```suggestion
    Fix a crash in
```

just minor - we don't categorise the "location" of bugfixes.  If we did, it'd be the `fixes:` header of the release note that changed (if you look in the template, you can see how we do that for features).

jakelishman reviewed (commented):
Thanks for the fix of the present issue.  Looking now at this pass, I think there's quite a lot more bugs in it:
- just above the visible diff in the review, there's a line like `for swap in dag.op_nodes(SwapGate)[::-1]:`.  That seems to be a false attempt to produce a reverse topological order - if the DAG was built backwards or had node replacements in the middle of it, it may not produce correct answers
- the check for swaps being "final" will probably keep re-checking the same structure all the time.  It might benefit from using Rustworkx's `TopologicalSorter(reverse=True, initial=list(dag.output_map.values())` to perform a true reverse-topological ordering from the back, and only mark nodes as "visited" if they're a measure or a swap, then we'll have a list that's a true reverse-topological order.
- This pass modifies the virtual-physical qubit assignment at the end of the circuit, so it should probably be updating the property set's `final_layout` to account for the modified permutation.

I'm not saying this PR needs to handle all or any of those, but if you're using this pass, is there any chance any of those might also be causing you trouble?  We can track in separate issues if so.

p51lee committed:
Clarify docstring in test_optimize_swap_2measure_same_clbit

p51lee commented:
Thanks—I agree on all three. Maybe my workloads wouldn’t be affected by them, but I’m happy to follow up. I’ll keep this PR to the crash fix. Do you prefer one tracking issue with a couple of small PRs, or separate issues?

p51lee committed:
Update releasenotes/notes/opt-swap-measure-clbit-reuse-938bab7e1968bea5.yaml

Co-authored-by: Jake Lishman <jake@binhbar.com>

p51lee committed:
Early-filter measure descendants in `OptimizeSwapBeforeMeasure`

Filter to Measure descendants first and only sort when >1 remain. This
avoids full-DAG scans in the common case (single measure) and keeps
behavior unchanged.

p51lee committed:
Merge branch 'main' into fix/opt-swap-measure-clbit-reuse

p51lee committed:
Merge branch 'main' into fix/opt-swap-measure-clbit-reuse

p51lee committed:
Merge branch 'main' into fix/opt-swap-measure-clbit-reuse
